mongo/integration/cmap_spec_test.go

```diff
@@ -0,0 +1,382 @@
+				ConnectionID: 1,
+			},
+			{Type: "ConnectionCreated"},
+			{
+				Type:         "ConnectionReady",
+				ConnectionID: 1,
+			},
+			{Type: "ConnectionCreated"},
+			{Type: "ConnectionReady"},
+			{Type: "ConnectionReady"},
+		}
+		ignore := []string{
+			"ConnectionCheckOutStarted",
+			"ConnectionCheckedIn",
+			"ConnectionCheckedOut",
+			"ConnectionClosed",
+			"ConnectionPoolCreated",
+			"ConnectionPoolReady",
+		}
+		assertEvents(mt, want, ignore, tpm.Events())
+	})
+
+	// Test that waiting on maxConnecting is limited by WaitQueueTimeoutMS.
+	// Based on CMAP spec test
+	// "connection-monitoring-and-pooling/pool-checkout-maxConnecting-timeout.json"
+	mt.Run("pool-checkout-maxConnecting-timeout.json", func(mt *mtest.T) {
+		mt.SetFailPoint(mtest.FailPoint{
+			ConfigureFailPoint: "failCommand",
+			Mode: mtest.FailPointMode{
+				Times: 50,
+			},
+			Data: mtest.FailPointData{
+				FailCommands:    []string{"hello", internal.LegacyHello},
+				CloseConnection: false,
+				BlockConnection: true,
+				BlockTimeMS:     750,
+			},
+		})
+
+		tpm := newTestPoolMonitor()
+		mt.ResetClient(options.Client().
+			SetPoolMonitor(tpm.PoolMonitor).
+			SetMaxPoolSize(10))
+
+		events := make(chan *event.PoolEvent, 20)
+		tpm.Subscribe(events)
+
+		var wg sync.WaitGroup
+		for i := 0; i < 2; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				_, err := mt.Coll.InsertOne(context.Background(), bson.D{{"key", "value"}})
+				assert.NoErrorf(mt, err, "InsertOne error")
+			}()
+		}
+
+		count := 2
+		for evt := range events {
+			if evt.Type == "ConnectionCreated" {
+				count--
+			}
+			if count <= 0 {
+				break
+			}
+		}
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			ctx, cancel := context.WithTimeout(context.Background(), 50*time.Millisecond)
+			defer cancel()
+			_, err := mt.Coll.InsertOne(ctx, bson.D{{"key", "value"}})
+			assert.IsTypef(
+				mt,
+				topology.WaitQueueTimeoutError{},
+				err,
+				"expected InsertOne error to be a WaitQueueTimeoutError")
+		}()
+
+		count = 1
+		for evt := range events {
+			if evt.Type == "ConnectionCheckOutFailed" {
+				count--
+			}
+			if count <= 0 {
+				break
+			}
+		}
+
+		wg.Wait()
+
+		want := []*event.PoolEvent{
+			{Type: "ConnectionCheckOutStarted"},
+			{Type: "ConnectionCheckOutStarted"},
+			{Type: "ConnectionCheckOutStarted"},
+			{
+				Type:   "ConnectionCheckOutFailed",
+				Reason: "timeout",
+			},
+		}
+		ignore := []string{
+			"ConnectionCreated",
+			"ConnectionCheckedIn",
+			"ConnectionCheckedOut",
+			"ConnectionClosed",
+			"ConnectionPoolCreated",
+			"ConnectionPoolReady",
+		}
+		assertEvents(mt, want, ignore, tpm.Events())
+	})
+
+	// Test that threads blocked by maxConnecting check out returned connections.
+	// Based on CMAP spec test
+	// "connection-monitoring-and-pooling/pool-checkout-returned-connection-maxConnecting.json"
+	// Note that the sequence of operations is modified from the original CMAP spec test to be able
```

Comment:
Nicely done. I think this exercises the same behavior. Using `BlockTimeMS` on `insert` to keep ready connections checked out is a good idea.